packages/react-native/Libraries/XR/RCTXRModule.mm

```diff
@@ -20,30 +21,52 @@ @implementation RCTXRModule {
                   : (RCTPromiseResolveBlock)resolve reject
                   : (RCTPromiseRejectBlock)reject)
 {
-  [self removeImmersiveBridge];
+  [self removeViewController:self->_immersiveBridgeView];
+  self->_immersiveBridgeView = nil;
+  RCTExecuteOnMainQueue(^{
+    if (self->_currentSessionId != nil) {
+      NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
+      [notificationCenter postNotificationName:@"RCTDismissImmersiveSpace" object:self userInfo:@{@"id": self->_currentSessionId }];
+    }
+  });
+  _currentSessionId = nil;
   resolve(nil);
 }
 
 
 RCT_EXPORT_METHOD(requestSession
-                  : (NSString *)sessionId resolve
-                  : (RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
+                  : (NSString *)sessionId userInfo
+                  : (NSDictionary *)userInfo resolve
+                  : (RCTPromiseResolveBlock)resolve reject
+                  : (RCTPromiseRejectBlock)reject)
 {
   RCTExecuteOnMainQueue(^{
     UIWindow *keyWindow = RCTKeyWindow();
     UIViewController *rootViewController = keyWindow.rootViewController;
     
     if (self->_immersiveBridgeView == nil) {
-      self->_immersiveBridgeView = [ImmersiveBridgeFactory makeImmersiveBridgeViewWithSpaceId:sessionId
+      NSMutableDictionary *userInfoDict = [[NSMutableDictionary alloc] init];
+      [userInfoDict setValue:sessionId forKey:@"id"];
+      if (userInfo != nil) {
+        [userInfoDict setValue:userInfo forKey:@"userInfo"];
+      }
+      NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
+          [notificationCenter postNotificationName:@"RCTOpenImmersiveSpace" object:self userInfo:userInfoDict];
+      self->_currentSessionId = sessionId;
+      
+      self->_immersiveBridgeView = [SwiftUIBridgeFactory makeImmersiveBridgeViewWithSpaceId:sessionId
                                                                             completionHandler:^(enum ImmersiveSpaceResult result){
         if (result == ImmersiveSpaceResultError) {
           reject(@"ERROR", @"Immersive Space failed to open, the system cannot fulfill the request.", nil);
-          [self removeImmersiveBridge];
+          [self removeViewController:self->_immersiveBridgeView];
+          self->_immersiveBridgeView = nil;
         } else if (result == ImmersiveSpaceResultUserCancelled) {
           reject(@"ERROR", @"Immersive Space canceled by user", nil);
-          [self removeImmersiveBridge];
+          [self removeViewController:self->_immersiveBridgeView];
+          self->_immersiveBridgeView = nil;
         } else if (result == ImmersiveSpaceResultOpened) {
-          resolve(nil);
+//          resolve(nil);
```

Comment:
Fix this as it crashes the app

Reply:
Won’t this leave a dangling promise if you won’t resolve it? Is something not checking for nil in the resolution chain?

Reply:
During development, I had an error that I've already resolved promise two times but after cleaning up everything and uncommenting it - it works fine now 👍🏻